src/util/alloc/allocator.rs

```diff
@@ -28,44 +28,124 @@ pub enum AllocationError {
     MmapOutOfMemory,
 }
 
-/// Behavior when an allocation fails, and a GC is expected.
-#[repr(u8)]
-#[derive(Copy, Clone, Default, PartialEq, bytemuck::NoUninit, Debug)]
-pub enum OnAllocationFail {
-    /// Request the GC. This is the default behavior.
-    #[default]
-    RequestGC,
-    /// Instead of requesting GC, the allocation request returns with a failure value.
-    ReturnFailure,
-    /// Instead of requesting GC, the allocation request simply overcommits the memory,
-    /// and return a valid result at its best efforts.
-    OverCommit,
+/// Allow specifying different behaviors with [`Allocator::alloc_with_options`].
+#[repr(C)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct AllocationOptions {
+    /// Should we poll *before* trying to acquire more pages from the page resource?
+    ///
+    /// **The default is `true`**.
+    ///
+    /// If `true`, the allocation will let the GC trigger poll before acquiring pages from the page
+    /// resource, giving the GC trigger a chance to schedule a collection.
+    ///
+    /// If `false`, the allocation will not notify the GC trigger *before* acquiring pages from the
+    /// page resource.  Note that if the allocation is at a safepoint (i.e. [`Self::at_safepoint`]
+    /// is true), it will still poll and force a GC *after* failing to get pages from the page
+    /// resource due to physical memory exhaustion.
+    pub eager_polling: bool,
```

Comment:
From users' point of view, they just call `alloc` and they don't see eager poll or lazy poll -- they either want their `alloc` call to poll or not poll. As we discussed, we can remove this option, or replace with `allow_polling` which affects both polling in `acquire`.